ukrainskyi-vitalii/nodejs-aws-shop-react-backend
# Task 10 (Backend For Frontend)

What was done?
1. Task 10.1
- BFF Service has been created
- request mapping according to .env file
- BFF Service returns the result of the recipient’s request
- If BFF Service cannot find the recipient URL, it returns a "Cannot process request" error message with status 502
- BFF Service returns the same status code and error message that the recipient service returns to the BFF Service in case of any error on the recipient service side.
2. Task 10.2
- BFF Service has been deployed with Elastic Beanstalk.
- BFF Service works only with requests from the Product Service and Cart Service.
- All Product Service and Cart Service methods should work correctly if requested via BFF Service
3. Done 11.08.2024 / Deadline 12.08.2024
4. Score: 100/100

Evaluation criteria (80 points for covering all criteria)

Provide your reviewers with the following information:

- [Link to the repo](https://github.com/ukrainskyi-vitalii/nodejs-aws-shop-react-backend/tree/task-10/bff-service)
- [Product Service API URL](https://nfewdukm7g.execute-api.eu-west-1.amazonaws.com/prod)
- [Cart Service API URL](https://ukrainskyi-vitalii-cart-api-dev.eu-west-1.elasticbeanstalk.com)
- [BFF Service API URL](http://ukrainskyi-vitalii-bff-api-test.eu-west-1.elasticbeanstalk.com/)
- example of how to call createProduct lambda with all needed information: URL, payload, headers, etc.
POST https://nfewdukm7g.execute-api.eu-west-1.amazonaws.com/prod/products
{"count": 1, "price": 10, "description": "Test Desc", "title": "Test Product"}

- example how to call Product Service and Cart Service via BFF Service URL

Get All Products
GET http://ukrainskyi-vitalii-bff-api-test.eu-west-1.elasticbeanstalk.com/product/products/
![image](https://github.com/user-attachments/assets/90be3001-aaf2-48d8-bb4c-27db261d66df)

Get Product By ID
GET http://ukrainskyi-vitalii-bff-api-test.eu-west-1.elasticbeanstalk.com/product/products/643e7cc1-5b4a-493f-b77b-e21635cea13a
![image](https://github.com/user-attachments/assets/d195de2d-3a26-433c-a6e4-9438b5924c91)

Get Cart
GET http://ukrainskyi-vitalii-bff-api-test.eu-west-1.elasticbeanstalk.com/cart/profile/cart
![image](https://github.com/user-attachments/assets/4b7919d6-eb8e-4568-ad55-c1d218b74f46)

Update Cart Item
PUT http://ukrainskyi-vitalii-bff-api-test.eu-west-1.elasticbeanstalk.com/cart/profile/cart
{"product":{"id":"8a24de54-5d25-4c73-a5f3-373f35509d10","description":" Handcrafted from genuine leather. This vintage-style journal is perfect for writing and preserving your thoughts. ","price":89,"title":"Vintage Leather Journal"},"count":2}
#Please keep the same json format in one line, otherwise the server triggers an error#
![image](https://github.com/user-attachments/assets/e4e8d959-f17e-40e9-9294-67ff40691402)

Create a New Order
PUT http://ukrainskyi-vitalii-bff-api-test.eu-west-1.elasticbeanstalk.com/cart/order
{"items":[{"productId":"8a24de54-5d25-4c73-a5f3-373f35509d10","count":1}],"address":{"comment":"comment","address":"street","lastName":"Test PUT Order","firstName":"first name"}}
![image](https://github.com/user-attachments/assets/2f37b913-85d2-4b62-9c83-57e2ad17f49e)

Get All Orders
GET http://ukrainskyi-vitalii-bff-api-test.eu-west-1.elasticbeanstalk.com/cart/order
![image](https://github.com/user-attachments/assets/61d5baf9-73f4-4444-8e14-f869db8ac782)

Get Order By ID
GET http://ukrainskyi-vitalii-bff-api-test.eu-west-1.elasticbeanstalk.com/cart/order/8c1600dd-416c-43c2-8ed0-062129e34501
![image](https://github.com/user-attachments/assets/b2faf0f9-6167-4733-8292-f874304e67aa)

Delete Order By ID
DELETE http://ukrainskyi-vitalii-bff-api-test.eu-west-1.elasticbeanstalk.com/cart/order/861c0b80-91d0-41f2-8d00-9c0732ce7111
![image](https://github.com/user-attachments/assets/946f0b48-ac50-411a-8232-44e814a572f3)

- A working and correct application should be in the bff-service folder. Reviewer can start this application locally with any valid configuration in the .env file and this application should works as described in the Task 10.1
[.env file](https://github.com/ukrainskyi-vitalii/nodejs-aws-shop-react-backend/blob/task-10/bff-service/.env)

- The BFF Service should be deployed with Elastic Beanstalk. The BFF Service call should be redirected to the appropriate service : Product Service or Cart Service. The response from the BFF Service should be the same as if Product Service or Cart Service services were called directly.
![image](https://github.com/user-attachments/assets/61428e63-4346-4ba8-8811-473fe7b963b2)


By this point your application must be able to do:
[FE application](https://d1zc7xy2yxaz04.cloudfront.net/)
[FE pull request](https://github.com/ukrainskyi-vitalii/nodejs-aws-shop-react-frontend/pull/6)

1. Products representation on Home page should be based on Product Service API.
2. Products are coming from Product DB.
3. Product images are not randomly generated on client side. Product image, same as another product model information
4. Should be stored on BE side in Product DB.
5. Products might be created through CSV product file import from client side.
6. Cart might be created with appropriate product set.
7. Auth logic should be in place
Note: product import API is covered with Auth logic

Additional (optional) tasks
+20 - Add a cache at the BFF Service level for a request to the getProductsList lambda function of the Product Service. The cache should expire in 2 minutes.
How to test:
- Get products list
GET http://ukrainskyi-vitalii-bff-api-test.eu-west-1.elasticbeanstalk.com/product/products
- Create new product
POST https://nfewdukm7g.execute-api.eu-west-1.amazonaws.com/prod/products
{"count": 1, "price": 10, "description": "Test Desc", "title": "Test Product Cache"}
- Get products list - result shouldn’t have new product
- Wait more than 2 minutes
- Get products list - result should have new product
